### PR TITLE
Add GPU support for extruded 1D spaces

### PR DESCRIPTION
**Pull request by dennisYatunin**
This PR adds GPU support for extruded 1D spaces, with the following modifications:
- Remove the assumption that `Ni == Nj` from `mapreduce_cuda_kernel`, `columnwise_partition`, `columnwise_universal_index`, and `multiple_field_solve_universal_index`
- Add methods for `operator_shmem` `operator_fill_shmem!`, and `operator_evaluate` corresponding to the 1D versions of all SEM operators
- Rename `DSSDataTypes` to `DSSTypes2D`, and define a new constant `DSSTypes1D`
- Add a method for `dss_1d!` corresponding to `CUDADevice`, and extend `dss!` to support both 1D and 2D data types
- Adapt the `local_geometry` and `dss_weights` of a `SpectralElementSpace1D` from `Array`s to `CuArray`s

These changes are also tested through CI:
- Add a CUDA job that runs `unit_2d.jl` and `convergence_2d.jl`
- Add tests for `IFH`, `IHF`, `VIFH`, and `VIHF`  to `unit_mapreduce.jl`, which already has a CUDA job

To fix several test failues, I had to make a few additional changes:
- Simplify the CUDA implementation of `Gradient` by directly differentiating vectors instead of breaking them up into components, as the original design could not distinguish the gradient of a scalar from the gradient of a single-component vector, and one of the tests in `unit_2d.jl` takes the gradient of a `UVector` field
- Extend `strip_space` to `FiniteDifferenceOperator` boundary conditions, as one of the tests in `unit_2d.jl` uses a 1D field for the `CurlC2F` boundary condition
- Reduce the number of methods for `level(space::ExtrudedFiniteDifferenceSpace, v)` from 4 to 1 to avoid a GPU inference failure in `reconstruct_placeholder_space` when evaluating boundary conditions (we hit some sort of compiler heuristic when we use 2 or 4 specialized methods, so we need to just have one method with union-splitting)
  - By making this change, I have inadvertently fixed the issue in #1989
- Adapt several `CuArray`s to `Array`s in `unit_2d.jl` and `unit_spaces.jl` to avoid scalar indexing errors
- Bump the number of allowed JET failures for `SpectralElementSpace1D` in `opt_spaces.jl` from 141 to 825 on GPUs and from 137 to 150 on CPUs

---

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: ext/cuda/remapping_distributed.jl [hunk 1]
```diff
@@ -130,9 +130,7 @@ function set_interpolated_values_kernel!(
         out[i, j, k] = 0
         for t in 1:Nq
             out[i, j, k] +=
-                I[i, t] *
-                I[i, s] *
-                (
+                I[i, t] * (
```
Comment (charleskawczynski) [inline ext/cuda/remapping_distributed.jl:133]: Was this a bug? Why aren't our unit tests catching this?

File: ext/cuda/remapping_distributed.jl [hunk 2]
```diff
@@ -275,9 +273,9 @@ function set_interpolated_values_kernel!(
 
         h = local_horiz_indices[i]
         out[i, k] = 0
-        for t in 1:Nq, s in 1:Nq
+        for t in 1:Nq
```
Comment (charleskawczynski) [inline ext/cuda/remapping_distributed.jl:276]: Again, is this a bug?

Comment (dennisYatunin) [reply ext/cuda/remapping_distributed.jl:276]: Yes, this is also an untested bug.

**Review by charleskawczynski — COMMENTED** [review]
> Bump the number of allowed JET failures for SpectralElementSpace1D in opt_spaces.jl from 141 to 825 on GPUs and from 137 to 150 on CPUs

This is alarming, without manifest changes. Would you mind taking a look at what is causing the spike in inferences? (I think that script makes it pretty easy to modify & see)

File: ext/cuda/remapping_distributed.jl (hunk 1 above)
Comment (charleskawczynski) [reply ext/cuda/remapping_distributed.jl:133]: Ok, good to know. Can we add a unit test in ClimaCore to exercise this?

File: ext/cuda/remapping_distributed.jl (hunk 2 above)
Comment (charleskawczynski) [reply ext/cuda/remapping_distributed.jl:276]: Also, cc @Sbozzolo

File: src/Hypsography/Hypsography.jl [hunk 3]
```diff
@@ -77,10 +77,12 @@ end
 Locate the levels by linear interpolation between the surface field and the top
 of the domain, using the method of [GalChen1975](@cite).
 """
-struct LinearAdaption{F <: Fields.Field} <: HypsographyAdaption
+struct LinearAdaption{F} <: HypsographyAdaption
     surface::F
 end
 
+strip_field(::LinearAdaption) = LinearAdaption(nothing)
```
Comment (charleskawczynski) [inline src/Hypsography/Hypsography.jl:84]: Shouldn't this be `strip_space`? And it should keep the field values

File: ext/cuda/remapping_distributed.jl (hunk 1 above)
Comment (dennisYatunin) [reply ext/cuda/remapping_distributed.jl:133]: Sure, that's easy to add. I only discovered this bug last night so I'll add some tests today.

File: src/Hypsography/Hypsography.jl (hunk 3 above)
Comment (dennisYatunin) [reply src/Hypsography/Hypsography.jl:84]: There is no need to keep the field values in this function because they are passed in separately through `ref_z_to_physical_z`. I think it is best to avoid passing duplicate field values through here. In general, the design of this section should probably be refactored to split the surface elevations from the adaption specification.

File: src/Topologies/dss.jl [hunk 4]
```diff
@@ -684,7 +677,14 @@ end
 
 Computed unweighted/pure DSS of `data`.
 """
-function dss!(data::DSSDataTypes, topology::Topology2D)
+function dss!(data::DSSTypes1D, topology::IntervalTopology)
+    Nij = DataLayouts.get_Nij(data)
+    length(parent(data)) == 0 && return nothing
```
Comment (charleskawczynski) [inline src/Topologies/dss.jl:682]: I think there might be more cases here that we can short circuit. What do you think?

File: test/Operators/unit_operators_examples.jl [hunk 5]
```diff
@@ -57,21 +57,12 @@ grid = Grids.ExtrudedFiniteDifferenceGrid(
...
-        @test_broken begin
-            @. ᶜgradᵥ(level_field + ᶠscalar_field)
-        end
-    else
-        @. ᶜgradᵥ(level_field + ᶠscalar_field)
-    end
-end
+using Test
+@testset "Example broadcast expression on GPUs" begin
+    @test !isnothing(@. ᶜgradᵥ(level_field + ᶠscalar_field))
```
Comment (charleskawczynski) [inline test/Operators/unit_operators_examples.jl:67]: 🎉, thank you!

File: test/Spaces/opt_spaces.jl [hunk 6]
```diff
@@ -35,7 +35,7 @@ end
 #! format: off
     if ClimaComms.device(context) isa ClimaComms.CUDADevice
         test_n_failures(91,   TU.PointSpace, context)
-        test_n_failures(141,  TU.SpectralElementSpace1D, context)
+        test_n_failures(825,  TU.SpectralElementSpace1D, context)
```
Comment (charleskawczynski) [inline test/Spaces/opt_spaces.jl:38]: Can you please look and see what this spike is from?

Comment (charleskawczynski) [reply test/Spaces/opt_spaces.jl:38]: We might be able to use a function barrier to substantially reduce it

Comment (dennisYatunin) [reply test/Spaces/opt_spaces.jl:38]: This spike is caused by the conversion of `Array`s to `CuArray`s, and it is similar to the spike in JET failures from CPUs to GPUs for `SpectralElementSpace2D`.

Comment (dennisYatunin) [reply test/Spaces/opt_spaces.jl:38]: If we need to reduce this for `SpectralElementSpace1D`, we should probably apply the same fix to `SpectralElementSpace2D`

File: src/Topologies/dss.jl (hunk 4 above)
Comment (dennisYatunin) [reply src/Topologies/dss.jl:682]: You mean like whether we are using GL or GLL quadrature? I think we're just assuming GLL quadrature everywhere outside of a few unit tests. Or did you have something else in mind?

**Comment by charleskawczynski** [comment]
> This is alarming, without manifest changes. Would you mind taking a look at what is causing the spike in inferences? (I think that script makes it pretty easy to modify & see)

It turns out that these are all false warnings, induced by our `get!` caching. Commenting out `get!` in spectral element grids and interval meshes shows no inference failures.

File: test/Spaces/opt_spaces.jl (hunk 6 above)
Comment (charleskawczynski) [reply test/Spaces/opt_spaces.jl:38]: It turns out that these are all false warnings, induced by our `get!` caching. Commenting out `get!` in spectral element grids and interval meshes shows no inference failures.

File: src/Topologies/dss.jl (hunk 4 above)
Comment (charleskawczynski) [reply src/Topologies/dss.jl:682]: Just checking `Nh > 1` should be good (+GLL?)

File: ext/cuda/remapping_distributed.jl (hunk 1 above)
Comment (Sbozzolo) [reply ext/cuda/remapping_distributed.jl:133]: There are unit tests for 1D extruded spaces, but they were skipped on GPU. There are https://github.com/CliMA/ClimaCore.jl/blob/a1c6470b8ea8f32143be190feea56e91749e9128/test/Remapping/distributed_remapping.jl#L46-L147.

File: ext/cuda/remapping_distributed.jl (hunk 2 above)
Comment (Sbozzolo) [reply ext/cuda/remapping_distributed.jl:276]: This should been tested here: https://github.com/CliMA/ClimaCore.jl/blob/a1c6470b8ea8f32143be190feea56e91749e9128/test/Remapping/distributed_remapping.jl#L666-L760

**Review by charleskawczynski — APPROVED** [review]
I chatted with @dennisYatunin about this PR. These changes look fine, and I approve once we add some unit tests exercising the new pieces/features.

**Comment by Sbozzolo** [comment]
In https://github.com/CliMA/ClimaCore.jl/pull/2176, I am adding tests for GPU for the CommonSpaces module. I excluded the `SliceXZ` case from those tests, so you should be able to turn it on (it if #2176 gets merged before this).

File: ext/cuda/remapping_distributed.jl (hunk 1 above)
Comment (dennisYatunin) [reply ext/cuda/remapping_distributed.jl:133]: This test has been reenabled on GPUs and appears to be passing now.

File: ext/cuda/remapping_distributed.jl (hunk 2 above)
Comment (dennisYatunin) [reply ext/cuda/remapping_distributed.jl:276]: I think you meant here:
https://github.com/CliMA/ClimaCore.jl/blob/a1c6470b8ea8f32143be190feea56e91749e9128/test/Remapping/distributed_remapping.jl#L866-L897

This test has been reenabled on GPUs and appears to be passing now.

File: src/Hypsography/Hypsography.jl (hunk 3 above)
Comment (dennisYatunin) [reply src/Hypsography/Hypsography.jl]: I have redesigned this to avoid passing the field values separately, instead using a lazy broadcast to pass them through the adaption. Keeping everything in the adaption seems more in line with the original intention of this data structure, and the new pattern is also more generalizable.

File: src/Topologies/dss.jl (hunk 4 above)
Comment (dennisYatunin) [reply src/Topologies/dss.jl]: We cannot optimize out DSS in a horizontally periodic space when `Nh == 1`, since the opposite sides of the single horizontal element are still collocated. We can only optimize it out when we are using GL instead of GLL, or when there is no data to DSS. To check the first condition, I have defined the function `Quadratures.requires_dss(quadrature_style(space))`. To check the second condition, I am now using `sizeof(eltype(data))`, which is somewhat clearer than the `length(parent(data))` that was originally here.
